lib/nx_zlib.c

```diff
@@ -512,72 +552,63 @@ nx_devp_t nx_open(int nx_id)
 	vas_handle = nx_function_begin(NX_FUNC_COMP_GZIP, nx_id);
 
 	if (!vas_handle) {
+		pthread_mutex_unlock(&saved_nx_devp_mutex);
 		prt_err("nx_function_begin failed, errno %d\n", errno);
 		free(nx_devp);
 		return NULL;
 	}
 
 	nx_devp->vas_handle = vas_handle;
-	nx_devp->open_cnt++;
+	nx_devp->open_cnt = 1;  /* newly allocated nx_devp, so single user */
+	nx_devp->creator_pid = my_pid;
+	nx_devp->use_cnt = 1; 	/* New device, so reset the counter so we know
+				   when it's time to stop reusing it. */
 
-	if (disable_saved_nx_devp == 0) {
-		/* we will reuse this handle */
-		pid_is_saved = 1;
-		saved_pid = my_pid;
-		saved_nx_devp = nx_devp;
-	}
+	/* We are about to replace the saved handle, so make sure to
+	   free the existing one if there are no active users. */
+	if (saved_nx_devp != NULL && saved_nx_devp->open_cnt == 0)
+		nx_device_free(saved_nx_devp);
+
+	/* we will reuse this handle */
+	saved_nx_devp = nx_devp;
```

Comment:
`saved_nx_devp` is a global variable.
Why are we reusing this handle when `nx_config.max_vas_reuse_count == 0`?
I suspect lines 567..573 should be protected inside `if (nx_config.max_vas_reuse_count > 0)`.

Reply:
@tuliom You're correct, we should not be saving the handle if reuse is not enabled. I have changed the PR the fix that.